### PR TITLE
feat(client): support custom starwhale version in runtime yaml

### PR DESCRIPTION
**Pull request by goldenxinxing**
## Description

- Support custom starwhale version in yaml.
- Support `git+https...` format with version.

Details:
1. The effective priority is: 
`environment.starwhale_version in yaml > starwhale dependency in requirements > current starwhale.`
2. Build with custom starwhale version
```
api_version: 1.1
configs:
  conda:
    channels:
      - conda-forge
dependencies:
  - pip:
      - numpy
  - requirements-sw-lock.txt
environment:
  arch: noarch
  os: ubuntu:20.04
  starwhale_version: git+https://github.com/star-whale/starwhale.git@main#subdirectory=client&setup_py=client/setup.py#egg=starwhale
mode: venv
name: simple-test
```
[asciicast](https://asciinema.org/a/g0kgCrcMMW9R7hqrnozNoCFlC)

3. Restore In Cloud(base entrypoint):
image

4. Restore In Standalone(venv)
image

5. Restore In Standalone(conda)
image

## Modules
- [ ] UI
- [ ] Controller
- [ ] Agent
- [x] Client
- [ ] Python-SDK
- [ ] Others

## Checklist
- [x] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: client/starwhale/core/runtime/model.py [hunk 1]
```diff
@@ -1357,7 +1358,9 @@ def _lock_environment(
         self._manifest["environment"].update(
             {
                 "lock": {
-                    "starwhale_version": self._detected_sw_version or STARWHALE_VERSION,
+                    "starwhale_version": swrt_config.environment.starwhale_version
+                    or self._detected_sw_version
+                    or STARWHALE_VERSION,
```
Comment (jialeicui) [inline client/starwhale/core/runtime/model.py:1363]: If installed using git, what version of the image will be used?

File: client/starwhale/utils/venv.py [hunk 2]
```diff
@@ -189,7 +190,6 @@ def venv_install_req(
         return
 
     venvdir = str(venvdir)
-    req = str(req)
```
Comment (tianweidut) [inline client/starwhale/utils/venv.py:192]: we need to convert req to str, for example: Path type.  https://github.com/star-whale/starwhale/blob/main/client/starwhale/core/runtime/model.py#L346

venv_install_req is a basic function for the most dependencies installation.

File: client/starwhale/utils/venv.py [hunk 3]
```diff
@@ -875,20 +875,20 @@ def install_starwhale(
...
         str(prefix_path / "bin" / "python3"), SW_PYPI_PKG_NAME
     )
     if _existed and not force:
         console.info(f"{SW_PYPI_PKG_NAME} has already be installed at {prefix_path}")
         return
 
+    if version == "" or version == SW_DEV_DUMMY_VERSION:
+        version = get_downloadable_sw_version()
+
+    req = [SW_PYPI_PKG_NAME]
```
Comment (tianweidut) [inline client/starwhale/utils/venv.py:888]: req is not needed to change type.

File: client/starwhale/utils/venv.py [hunk 4]
```diff
@@ -875,20 +875,20 @@ def install_starwhale(
...
+    if version:
+        req = [version] if re.match(r"^git", version) else [f"{req}=={version}"]
```
Comment (tianweidut) [inline client/starwhale/utils/venv.py:890]: `[f"{req}=={version}"]`  typo?

req is a list type.  add some ut?

Comment (goldenxinxing) [reply client/starwhale/utils/venv.py]: fixed

File: client/starwhale/utils/venv.py (hunk 3 above)
Comment (goldenxinxing) [reply client/starwhale/utils/venv.py:888]: There will be an error `starwhale/utils/venv.py: error: Incompatible types in assignment (expression has type "List[str]", variable has type "str")` when lint with mypy

Comment (goldenxinxing) [reply client/starwhale/utils/venv.py]: fixed

**Review by tianweidut — APPROVED** [review]
LGTM

**Pull request merged**